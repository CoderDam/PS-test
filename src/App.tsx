import { ReactElement, useRef, useState } from 'react'
import {
  Box,
  Container,
  Fade,
  Grid,
  Grow,
  IconButton,
  Theme,
  Typography,
  useMediaQuery,
  useTheme
} from '@mui/material'
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward'

import Books from './components/Books'
import Cart from './components/Cart'
import Offers from './components/Offers'
import { BookType } from './types'

const boxStyle = (theme: Theme): any => ({
  background: 'linear-gradient(to right, #1e0c0c, #6f4e37 7% 93%, #1e0c0c)',
  display: 'flex',
  flexDirection: 'column',
  height: '100vh',
  overflowX: 'hidden',
  // keep the scrollbar visible to avoid UI jumps
  overflowY: 'scroll',
  // ...but make it discreet, on Firefox
  scrollbarWidth: 'thin',
  scrollbarColor: '#4b3621 #1e0c0c',
  // ...and Chrome, Edge, Safari
  '::-webkit-scrollbar': {
    width: 8
  },
  '::-webkit-scrollbar-track': {
    background: '#1e0c0c'
  },
  '::-webkit-scrollbar-thumb': {
    background: '#4b3621',
    borderRadius: 8
  }
})
const buttonContainerStyle = (theme: Theme): any => ({
  bottom: theme.spacing(3),
  left: '50vw',
  position: 'fixed',
  transform: 'translateX(-50%)',
  zIndex: 20
})
const buttonStyle = (theme: Theme): any => ({
  backgroundColor: theme.palette.primary.light,
  ':active, :hover': {
    backgroundColor: theme.palette.primary.dark
  }
})

function App (): ReactElement {
  const [selectedBooks, setSelectedBooks] = useState<BookType[]>([])
  const theme = useTheme()
  const isDownMd = useMediaQuery(theme.breakpoints.down('md'))
  const boxRef = useRef<HTMLElement>(null)

  const onScrollBottom = (): void => boxRef.current?.scrollTo?.({ top: 10000, behavior: 'smooth' })
  const onAddBook = (book: BookType): void => setSelectedBooks((prevBooks) => [...prevBooks, book])
  const onDeleteBook = (book: BookType): void =>
    setSelectedBooks((prevBooks) => prevBooks.filter((prevBook) => prevBook.isbn !== book.isbn))

  return (
    <Box sx={boxStyle} ref={boxRef}>
      <Container
        sx={{
          backgroundColor: isDownMd ? '#6f4e37' : 'transparent',
          flex: '1 0 auto',
          paddingBottom: theme.spacing(4),
          paddingTop: theme.spacing(2)
        }}
      >
        <Typography
          variant='h4'
          align='center'
          color='primary'
          paragraph
          sx={{ padding: theme.spacing(0, 2) }}
        >
          Bienvenue dans la biblioth√®que d'
          <strong>Henry Potier</strong>
        </Typography>
        <Grid container spacing={3}>
          <Grid item xs={12} md={7} sx={{ display: 'flex' }}>
            <Books
              selectedBooks={selectedBooks}
              onAddBook={onAddBook}
              onDeleteBook={onDeleteBook}
            />
          </Grid>
          {selectedBooks.length > 0 && (
            <Grid
              item
              xs={12}
              md={5}
              sx={{ display: 'flex', paddingBottom: (theme) => theme.spacing(3) }}
            >
              <Cart selectedBooks={selectedBooks} onDeleteBook={onDeleteBook} />
            </Grid>
          )}
        </Grid>
        {selectedBooks.length > 1 && (
          <Fade in>
            <Offers selectedBooks={selectedBooks} />
          </Fade>
        )}
      </Container>
      {isDownMd && selectedBooks.length > 0 && (
        <Box sx={buttonContainerStyle}>
          <Grow in>
            <IconButton size='large' sx={buttonStyle} onClick={onScrollBottom}>
              <ArrowDownwardIcon />
            </IconButton>
          </Grow>
        </Box>
      )}
    </Box>
  )
}

export default App
