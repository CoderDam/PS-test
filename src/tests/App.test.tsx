import { render, screen } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from 'react-query'

import App from '../App'

const queryClient = new QueryClient()

describe('App', () => {
  test('renders "bienvenue"', () => {
    const app = (
      <QueryClientProvider client={queryClient}>
        <App />
      </QueryClientProvider>
    )
    render(app)
    const title = screen.getByText(/bienvenue/i)
    expect(title).toBeInTheDocument()
  })
  test('snapshot', () => {
    const app = (
      <QueryClientProvider client={queryClient}>
        <App />
      </QueryClientProvider>
    )
    const { asFragment } = render(app)
    expect(asFragment()).toMatchSnapshot()
  })
})
