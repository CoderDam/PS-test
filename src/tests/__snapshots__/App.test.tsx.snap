// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-yj71k8"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1bkqgu-MuiContainer-root"
    >
      <p
        class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-paragraph css-1rwpz78-MuiTypography-root"
      >
        Bienvenue dans la biblioth√®que d'
        <strong>
          Henry Potier
        </strong>
      </p>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-7 css-tjynou-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-1qi6aoa"
          >
            <p
              class="MuiTypography-root MuiTypography-h5 MuiTypography-paragraph css-tidlkh-MuiTypography-root"
            >
              Voici les livres que vous pouvez consulter ici :
            </p>
            <div
              class="css-obcb4a"
            >
              <div
                class="bookshelf_container"
                style="transform: scale(1); margin-bottom: 0%;"
              >
                <div
                  class="bookshelf_wrapper"
                >
                  <ul
                    class="books_list"
                  >
                    <li
                      class="book_item first"
                    />
                    <li
                      class="book_item second"
                    />
                    <li
                      class="book_item third"
                    />
                    <li
                      class="book_item fourth"
                    />
                    <li
                      class="book_item fifth"
                    />
                    <li
                      class="book_item sixth"
                    />
                  </ul>
                  <div
                    class="shelf"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
